### DISCUSSION TITLE
Secrets for PRs who are not collaborators

### DISCUSSION THREAD

**Discussion by kaihendry** [community/community#26163] [Actions]
<p>I want to be able to preview on Netlify the PRs that outside contributors propose.</p>
<p>As you can see from <a href="https://github.com/nwtgck/actions-netlify/issues/32#issuecomment-591800758" rel="nofollow noopener">https://github.com/nwtgck/actions-netlify/issues/32#issuecomment-591800758</a></p>
<p>Secrets are not passed to workflows that are triggered by a pull request from a fork.</p>
<p>Is there some sane way to work around this?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/kaihendry">@kaihendry</a> ,</p>
<p>Yeah, as you have saw from the docs, <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#using-encrypted-secrets-in-a-workflow" rel="nofollow noopener"><strong>secrets are not passed to the runner when a workflow is triggered from a forked repository.</strong></a> This is for security reasons, and currently we have no any workaround to bypass this limitation.</p>
<p>Of course, if your projects really need this feature, I recommend you directly report your feature request <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">here</a>. That will allow you to directly interact with the appropriate engineering team, and make it more convenient for the engineering team to collect and categorize your suggestions.</p>

**Comment by kaihendry** [comment]
<p>How do I express that I build on PR but only if it is a collaborator to avoid the failure messages?</p>
<pre><code>on:
    pull_request:
      types: [opened, synchronize]
</code></pre>
<p>Is what I have in <a href="https://github.com/kaihendry/dabase.com/blob/ec94d0b8de9b12f04e1428e1f02ce0a1787dfbf0/.github/workflows/netlify.yml" rel="nofollow noopener">https://github.com/kaihendry/dabase.com/blob/ec94d0b8de9b12f04e1428e1f02ce0a1787dfbf0/.github/workflows/netlify.yml</a></p>
<p>Or perhaps can I bless a PR to built under my user? For the preview I want?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/kaihendry">@kaihendry</a> ,</p>
<p>I have a workaround, maybe you can reference:</p>
<ol>
<li>
<p>In your original repository, create a branch bashed on the default/base branch, for example create <strong>Merge_Fork</strong> branch based on master branch.</p>
</li>
<li>
<p>In the forked repository, every time the collaborators commit some new changes, ask the collaborators create a PR to merge the new changes from the forked repository into the Merge_Fork branch of the original repository. Set this PR do not trigger the workflow (using branch filter).</p>
</li>
<li>
<p>In the original repository, create a PR to merge teh changes from Merge_Fork branch into master branch, set this PR trigger the workflow (using branch filter), in this situation the workflow can access secrets.</p>
<p>on:<br>
pull_request:<br>
types: [opened, synchronize]<br>
branches:<br>
- master</p>
</li>
</ol>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/kaihendry">@kaihendry</a> ,</p>
<p>As I suggested in previous reply, create a feature branch in your original repository. At first merge changes from the forked repository into the feature branch of the original repository, and then create a PR to merge changes from the feature branch to the base branch to trigger the workflow run, this will allow the workflow to access the secrets in your original repository.</p>
<p>Have you tried this suggestion? Is it helpful to you? Any progress, please feel free to let me know.</p>

**Comment by kaihendry** [comment]
<p>There is no automated way to do this right?</p>
<p>The whole point of using Github and this PR flow is to make things easier. If I have to run a bunch of commands then I feel the value of this flow is diminished.</p>

**Comment by nwtgck** [comment]
<p>Although this is just an idea, that automcation may be created with GitHub Actions because GITHUB_TOKEN has write permission to original repo.</p>
<p>(I try to make this proof of concept.)</p>

**Comment by nwtgck** [comment]
<p>I made a PoC GitHub Actions for merge preview.</p>
<p>Here is a successfull merge preview.</p>
<p><a href="https://github.com/nwtgck/actions-merge-preview/runs/489448436?check_suite_focus=true" rel="nofollow noopener">https://github.com/nwtgck/actions-merge-preview/runs/489448436?check_suite_focus=true</a></p>
<p>My expected usage is that owner comments “<span class="mention">@some-bot</span> merge preview”,  then the actions is triggered by the comment in PR.</p>

**Comment by kaihendry** [comment]
<p>Ah, this looks promising. But I am not sure how to view your workflow file.</p>
<p>operation-test just appears to do a check out?</p>

**Comment by nwtgck** [comment]
<p>Thanks.</p>
<p>I made a demo video to tell how to use.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/da9a1a9108c5ea8fbe84fb3537a5e773e3dde9eb.gif" data-download-href="/uploads/short-url/vbQfLOw3wlNwxEmCXKf7MX9OINZ.gif?dl=1" title="actions-merge-preview.gif" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/10933561/181095882-23ae3492-81c2-451e-92c5-bd8f15c3334a.gif" alt="actions-merge-preview.gif" data-base62-sha1="vbQfLOw3wlNwxEmCXKf7MX9OINZ" width="690" height="388" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/d/da9a1a9108c5ea8fbe84fb3537a5e773e3dde9eb_2_690x388.gif, https://aws1.discourse-cdn.com/github/original/2X/d/da9a1a9108c5ea8fbe84fb3537a5e773e3dde9eb.gif 1.5x, https://aws1.discourse-cdn.com/github/original/2X/d/da9a1a9108c5ea8fbe84fb3537a5e773e3dde9eb.gif 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/da9a1a9108c5ea8fbe84fb3537a5e773e3dde9eb_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">actions-merge-preview.gif</span><span class="informations">700×394 1.12 MB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Here is an actual pull request from non-maintainers:<br>
<a href="https://github.com/nwtgck/piping-chat-web/pull/194" rel="nofollow noopener">https://github.com/nwtgck/piping-chat-web/pull/194</a></p>
<p>Here is the workflow file:</p>
<p><a href="https://github.com/nwtgck/piping-chat-web/blob/762cd1541d3a12eb94849d8017e435e5ddedf992/.github/workflows/merge-preview.yml" rel="nofollow noopener">https://github.com/nwtgck/piping-chat-web/blob/762cd1541d3a12eb94849d8017e435e5ddedf992/.github/workflows/merge-preview.yml</a></p>

**Comment by nwtgck** [comment]
<p>I made another more general GitHub Actions than before. You will get preview-branch to run GitHub Actions with the secrets for forked repos.</p>
<p><img src="https://github.com/nwtgck/actions-comment-run/raw/20297f070391450752be7ac1ebd454fb53f62795/doc_assets/merge-preview.png" alt=""></p>
<p>You can execute arbitrary script on GitHub comment. Here is the code: <a href="https://github.com/nwtgck/actions-comment-run/tree/20297f070391450752be7ac1ebd454fb53f62795#pr-merge-preview" rel="nofollow noopener">https://github.com/nwtgck/actions-comment-run/tree/20297f070391450752be7ac1ebd454fb53f62795#pr-merge-preview</a>.</p>

**Comment by kaihendry** [comment]
<p>Aha! I understand…  only owner can run it <code>allowed-associations: '["OWNER"]'</code></p>
<p><a href="https://github.com/kaihendry/dabase.com/pull/7" rel="nofollow noopener">https://github.com/kaihendry/dabase.com/pull/7</a></p>
<p><strong>Smart!</strong></p>
<p>But I do lament copying in all that JS…</p>
<p>Couldn’t it be like a one click button to merge the PR into a local branch I wonder? Or some one-liner?</p>

**Comment by nwtgck** [comment, marked as answer]
<p>Thanks!</p>
<p>How about “Saved replies”? You can quickly use presaved comments.</p>
<p><img src="https://github.com/nwtgck/actions-comment-run/raw/20297f070391450752be7ac1ebd454fb53f62795/doc_assets/saved-replies.gif" alt="Saved replies"></p>

**Comment by kaihendry** [comment]
<p>Whoa, this makes Github actions <strong>next level</strong> in my opinion.</p>
<p>M$ should be sending you a cheque. :wink:</p>
